### PR TITLE
[HLSL] Implement `rsqrt` intrinsic

### LINKED ISSUE #70074: [HLSL] implement rsqrt intrinsic
Implement HLSL `rsqrt` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-rsqrt

### PR DESCRIPTION
**Pull request by farzonl**
This change implements #70074
- `hlsl_intrinsics.h` - add the `rsqrt` api
- `DXIL.td` add the llvm intrinsic to DXIL op lowering map.
- `Builtins.td` - add an hlsl builtin for rsqrt.
- `CGBuiltin.cpp` add the ir generation for the rsqrt intrinsic.
- `SemaChecking.cpp` - reuse the one arg float only  checks.
- `IntrinsicsDirectX.td` -add an `rsqrt` intrinsic.

### CODE REVIEW

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 1]
```diff
@@ -1153,6 +1153,38 @@ double3 rcp(double3);
 _HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_rcp)
 double4 rcp(double4);
 
+//===----------------------------------------------------------------------===//
+// rsqrt builtins
+//===----------------------------------------------------------------------===//
+
+/// \fn T rsqrt(T x)
+/// \brief RReturns the reciprocal of the square root of the specified value \a
```
Comment (damyanp) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:1161]: `Rreturns` - typo?

Comment (damyanp) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:1161]: Comparing this with the doc for rcp above, wonder if it should be more like:

`Returns the reciprocal of the square root of the specified value.  ie 1 / sqrt(\a x).`

File: clang/test/SemaHLSL/BuiltIns/dot-warning.ll [hunk 2]
```diff
@@ -0,0 +1,49 @@
+; ModuleID = 'D:\projects\llvm-project\clang\test\SemaHLSL\BuiltIns\dot-warning.hlsl'
```
Comment (damyanp) [inline clang/test/SemaHLSL/BuiltIns/dot-warning.ll:1]: Maybe naive question...but isn't this about the `dot` intrinsic?

File: clang/test/SemaHLSL/BuiltIns/rsqrt-errors.hlsl [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+float2 test_too_many_arg(float2 p0) {
+  return __builtin_hlsl_elementwise_rsqrt(p0, p0);
+  // expected-error@-1 {{too many arguments to function call, expected 1, have 2}}
+}
+
+float builtin_bool_to_float_type_promotion(bool p1) {
+  return __builtin_hlsl_elementwise_rsqrt(p1);
+  // expected-error@-1 {{1st argument must be a vector, integer or floating point type (was 'bool')}}
```
Comment (damyanp) [inline clang/test/SemaHLSL/BuiltIns/rsqrt-errors.hlsl:16]: "integer" - is this true?  I thought it only worked with things that were float-like.

File: clang/test/SemaHLSL/BuiltIns/dot-warning.ll (hunk 2 above)
Comment (farzonl) [reply clang/test/SemaHLSL/BuiltIns/dot-warning.ll:1]: this file shouldn't be in this pr. this was a mistake.

**Review by damyanp — APPROVED** [review]

**Review by bogner — APPROVED** [review]

**Review by bharadwajy — APPROVED** [review]

**Pull request merged**